### ISSUE TITLE
Logically identical code. One compiles, one does not

### ISSUE THREAD

**Issue by schneems** [rust-lang/rust#116572]
I tried to compile this code [playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3a7b73546d419d316ad792d0c6362152):

```
use std::time::{Duration, Instant};
use std::thread;

fn time_bounded_retry<T, E>(
    max_time: Duration,
    sleep_for: Duration,
    f: impl Fn() -> Result<T, E>,
) -> Result<T, E> {
    let start = Instant::now();
    while let result = f() {
        match result {
            Ok(_) => return result,
            Err(_) => {
                if start.elapsed() > max_time {
                    return result;
                }

                thread::sleep(sleep_for);
            }
        }
    }
    // unreachable!()
}
```

However, it produces an error. If you uncomment the `unreachable!()` will work. Here's the error message:

```
    |
191 |     while let result = f() {
    |     ^^^^^^^^^^^^^^^^^^^^^^ this might have zero elements to iterate on
192 |         match result {
193 |             Ok(_) => return result,
    |                      ------------- if the loop doesn't execute, this value would never get returned
...
196 |                     return result;
    |                     ------------- if the loop doesn't execute, this value would never get returned
    = help: return a value for the case when the loop has zero elements to iterate on, or consider changing the return type to account for that possibility
help: try adding an expression at the end of the block
```

This code is functionally identical to the `while let` code and can compile without an `unreachable!()`:

```
fn time_bounded_retry<T, E>(
    max_time: Duration,
    sleep_for: Duration,
    f: impl Fn() -> Result<T, E>,
) -> Result<T, E> {
    let start = Instant::now();
    loop {
        let result = f();
        match result {
            Ok(_) => return result,
            Err(_) => {
                if start.elapsed() > max_time {
                    return result;
                }

                thread::sleep(sleep_for);
            }
        }
    }
}
```

The code in the while loop is irrefutable, so it is functionally the same as the `loop` case. Even if it's not a good practice to use an irrefutable statement in a while loop (and clippy will catch this) I think it should still compile.

My main reason for wanting it to compile is that while iterating, I wrote the `while let` version first, and because I could not get it to compile without the `unreachable!()` statement mistakenly believed that the compiler would not allow something like the second result. As a result of that mistaken impression ended up with this code:

```
fn time_bounded_retry<T, E>(
    max_time: Duration,
    sleep_for: Duration,
    f: impl Fn() -> Result<T, E>,
) -> Result<T, E> {
    let start = Instant::now();
    let mut result;
    loop {
        result = f();
        if result.is_ok() || start.elapsed() > max_time {
            break;
        }
        thread::sleep(sleep_for);
    }
    result
}
```

Which is not ideal (needs a temp mutable variable). A co-worker without the above learned helplessness `unreachable!()` experience wrote this code below, which I believe is a much better way to implement the logic as it does not require a temp mutable variable:

```
fn time_bounded_retry<T, E>(
    max_time: Duration,
    sleep_for: Duration,
    f: impl Fn() -> Result<T, E>,
) -> Result<T, E> {
    let start = Instant::now();
    loop {
        result = f();
        if result.is_ok() || start.elapsed() > max_time {
            return result;
        }
        thread::sleep(sleep_for);
    }
}
```

So effectively, I'm advocating for `while let var = //...` to be treated the same as `loop { let var = //...`, not because I think it's a good idea, but because having one error and not the other gives false feedback to the coder.

### Meta

`rustc --version --verbose`:
```
$ rustc --version --verbose
rustc 1.73.0 (cc66ad468 2023-10-03)
binary: rustc
commit-hash: cc66ad468955717ab92600c770da8c1601a4ff33
commit-date: 2023-10-03
host: x86_64-apple-darwin
release: 1.73.0
LLVM version: 17.0.2
```

**Comment by Nadrieril** [comment]
If we do this, I'd expect `while let (x, y) = foo() {}` to also typecheck like a `loop {}`. This would require typechecking to depend on the code that checks refutability of patterns, which would be pretty hard to do today.

A workable solution would be instead to change the error message: the code that emits "this might have zero elements to iterate on" could check if the pattern is a single binding and emit a better error message if that's the case. I expect this is rather easy to do, so marking this as an easy issue.

**Comment by yossydev** [comment]
I would like to work on this issue, is it possible for you to assign me?

**Comment by Nadrieril** [comment]
You figured it out 👍 Open a topic on [#t-compiler/help on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp) when you have questions!

**Comment by yossydev** [comment]
@Nadrieril 
> You figured it out 👍 Open a topic on [#t-compiler/help on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp) when you have questions!

Thank you for your kind attention!

**Comment by maxprogrammer007** [comment]
In the while let version, the compiler sees the while let result = f() statement and tries to match the result of f() to the pattern result. However, since f() returns a Result<T, E> which can be either Ok(T) or Err(E), the compiler is unsure how to match this result to the pattern result. This ambiguity leads to the error you encountered.

On the other hand, in the loop version, the compiler doesn't have to match the result of f() to any pattern. It simply calls f() in each iteration and checks the result directly. This approach avoids the ambiguity and compiles successfully.

**Comment by Kohei316** [comment]
@Nadrieril
Is this still open? Can I work on it?

**Comment by Nadrieril** [comment]
Yes, go ahead!

**Comment by traviscross** [comment]
Minimization:

```rust
fn foo() -> bool {
    while let x = false {
        if x { return true }
    }
    //~^ ERROR mismatched types
}
```

Perhaps the compiler treats it similarly to this, which is what I would expect it might do:

```rust
fn foo() -> bool {
    loop {
        let x = false else { break };
        //                   ^^^^^
        //~^ ERROR mismatched types
        if x { return true }
    }
}
```

@rustbot labels +A-diagnostics

@Kohei316: To confirm, you're planning to improve the diagnostics as described above, not try to change the behavior, yes?

**Comment by Kohei316** [comment]
@traviscross 

> @Kohei316: To confirm, you're planning to improve the diagnostics as described above, not try to change the behavior, yes?

Yes.And thank you for your minimum code.It became clearer what to do.

**Comment by anoshyn** [comment]
Hi @Kohei316 — are you still planning to work on this? It's been a while since the issue was last active, so I wanted to check in before stepping on any in-flight work. If you've moved on or it's no longer on your plate, I'd be happy to take it over.

**Comment by anoshyn** [comment]
Quick triage note before committing — on a stage1 build of current `main` (`rustc 1.97.0-dev`), @traviscross's minimization now produces:

```text
error[E0308]: mismatched types
 --> src/lib.rs:2:5
  |
1 |   fn foo() -> bool {
  |               ---- expected `bool` because of return type
2 | /     while let x = false {
3 | |         if x { return true }
4 | |     }
  | |_____^ expected `bool`, found `()`
  |
  = note: `while` loops evaluate to unit type `()`
help: consider returning a value here
  |
4 ~     }
5 +     /* `bool` value */
  |
```

The misleading "this might have zero elements to iterate on" / "if the loop doesn't execute" framing from the original 2023 report is gone — the diagnostic now correctly explains that `while` loops evaluate to `()` and presents a normal return-type mismatch.

What still seems off: `while let x = false` is irrefutable, so the loop is effectively `loop { let x = …; … }`, the function body is diverging, and the suggestion to add a tail expression is misleading (there's no execution path to it). But addressing that appears to require detecting irrefutable `while let` patterns during type-checking — which feels more substantial than typical E-easy work to me.

Two quick questions before I dig further:
1. Should the `E-easy` label be reconsidered given the current state?
2. Or has the original concern been addressed enough to close this in favor of a narrower follow-up?

Either answer is useful — just want to make sure I'm scoped correctly before claiming.

**Comment by Nadrieril** [comment]
Detecting irrefutable patterns during typechecking properly is not really feasible, however for diagnostic purposes detecting the simple case of a `while let binding = ...` or similar shouldn't be hard. There's already a warning that does that properly, it's just that it doesn't get displayed because of the type error; I think adding a note about that to the error would be helpful.

**Comment by anoshyn** [comment]
Thanks @Nadrieril — that's clear.

It looks like the natural place is in `rustc_hir_typeck::coercion::CoerceMany::coerce_inner`, right where there's already a branch that adds the note about `while` loops evaluating to unit. The body is desugared to `if let pat = scrut { body } else { break; }`, so checking the pattern there avoids reaching into mir_build at all.

One scope question before I claim: which patterns should the new note cover?

- Just the obvious cases — `PatKind::Binding(_, _, _, None)` and `PatKind::Wild`?
- Or also tuples of those (like your `while let (x, y) = foo()` example)?

Happy to start with the first and leave anything fancier for a follow-up if that matches what you had in mind.

**Comment by Nadrieril** [comment]
I'll let you be the judge, either seems fine

**Comment by schneems** [comment]
Thank you!
